lib/ansible/plugins/connection/ssh.py

```diff
@@ -641,18 +645,49 @@ def put_file(self, in_path, out_path):
         # accept them for hostnames and IPv4 addresses too.
         host = '[%s]' % self.host
 
-        if C.DEFAULT_SCP_IF_SSH:
-            cmd = self._build_command('scp', in_path, u'{0}:{1}'.format(host, pipes.quote(out_path)))
-            in_data = None
-        else:
-            cmd = self._build_command('sftp', to_bytes(host))
-            in_data = u"put {0} {1}\n".format(pipes.quote(in_path), pipes.quote(out_path))
-
-        in_data = to_bytes(in_data, nonstring='passthru')
-        (returncode, stdout, stderr) = self._run(cmd, in_data)
-
-        if returncode != 0:
-            raise AnsibleError("failed to transfer file to {0}:\n{1}\n{2}".format(to_native(out_path), to_native(stdout), to_native(stderr)))
+        # since this can be a non-bool now, we need to handle it correctly
+        scp_if_ssh = C.DEFAULT_SCP_IF_SSH
+        if not isinstance(scp_if_ssh, bool):
+            scp_if_ssh = scp_if_ssh.lower()
+            if scp_if_ssh in BOOLEANS:
+                scp_if_ssh = boolean(scp_if_ssh)
+            elif scp_if_ssh != 'smart':
+                raise AnsibleOptionsError('scp_if_ssh needs to be one of [smart|True|False]')
+
+        # create a list of commands to use based on config options
+        methods = ['sftp']
+        if scp_if_ssh == 'smart':
+            methods.append('scp')
+        elif scp_if_ssh:
+            methods = ['scp']
+
+        success = False
+        res = None
+        for method in methods:
+            if method == 'sftp':
+                cmd = self._build_command('sftp', to_bytes(host))
+                in_data = u"put {0} {1}\n".format(pipes.quote(in_path), pipes.quote(out_path))
+            elif method == 'scp':
+                cmd = self._build_command('scp', in_path, u'{0}:{1}'.format(host, pipes.quote(out_path)))
+                in_data = None
+
```

Comment:
may be cleaner to build the list of methods, iterate on that list to build a list of (cmd,in_data,checkrc) for each method, then iterate over that list when invoking it. something like:

``` python
method_names =['sftp']
if scp_if_ssh == 'smart':
     method_names.append('scp')
elif scp_if_ssh:
     method_names = ['scp']

methods = []
for method_name in method_names:
    if method_name == 'sftp':
          methods.append(self._build_sftp_put_command(host, in_path, out_path))
   if method_name == 'scp':
         methods.append(self._build_scp_put_command(host, in_path, out_path))

for cmd, in_data, checkrc, method in methods:
    (returncode, stdout, stderr) = self._run(cmd, in_data, checkrc)
    if returncode == 0:
         return
   if scp_if_ssh == 'smart':
         <the display warning/debug code here>

# nothing succeeded and returned
raise AnsibleError...
```